File: test-support/src/main/java/com/dbx/agent/test/JdbcFakeExecutionBehaviorTest.java
package com.dbx.agent.test;

import com.dbx.agent.DatabaseAgent;
import com.dbx.agent.ExecuteQueryOptions;
import com.dbx.agent.QueryResult;
import org.junit.jupiter.api.Test;

import java.util.Collections;

import static org.junit.jupiter.api.Assertions.assertEquals;

public abstract class JdbcFakeExecutionBehaviorTest {
    protected abstract DatabaseAgent createAgent();

    protected abstract String resultSetSql();

    @Test
    void executesNonSelectStatementsThatReturnResultSets() {
        DatabaseAgent agent = createAgent();
        TestSupport.setPrivateConnection(agent, JdbcAgentFake.connection());

        QueryResult result = agent.executeQuery(resultSetSql(), null, new ExecuteQueryOptions());

        assertEquals(Collections.singletonList("VALUE"), result.getColumns());
        assertEquals(Collections.singletonList(Collections.singletonList("row-value")), result.getRows());
        assertEquals(asList("setMaxRows:10001", "execute"), JdbcAgentFake.calls);
    }

    @Test
    void appliesExecutionRowAndFetchLimitsToJdbcStatements() {
        DatabaseAgent agent = createAgent();
        TestSupport.setPrivateConnection(agent, JdbcAgentFake.connection());

        QueryResult result = agent.executeQuery(resultSetSql(), null, new ExecuteQueryOptions(1, 25));

        assertEquals(Collections.singletonList(Collections.singletonList("row-value")), result.getRows());
        assertEquals(asList("setMaxRows:2", "setFetchSize:25", "execute"), JdbcAgentFake.calls);
    }

    private static java.util.List<String> asList(String first, String second) {
        return java.util.Arrays.asList(first, second);
    }

    private static java.util.List<String> asList(String first, String second, String third) {
        return java.util.Arrays.asList(first, second, third);
    }
}

File: test-support/src/main/kotlin/com/dbx/agent/test/JdbcFakeExecutionBehaviorTest.kt
package com.dbx.agent.test

import com.dbx.agent.DatabaseAgent
import com.dbx.agent.ExecuteQueryOptions
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

abstract class JdbcFakeExecutionBehaviorTest {
    protected abstract fun createAgent(): DatabaseAgent

    protected abstract fun resultSetSql(): String

    @Test
    fun `executes non select statements that return result sets`() {
        val agent = createAgent()
        setPrivateConnection(agent, JdbcAgentFake.connection())

        val result = agent.executeQuery(resultSetSql(), null)

        assertEquals(listOf("VALUE"), result.columns)
        assertEquals(listOf(listOf("row-value")), result.rows)
        assertEquals(listOf("setMaxRows:10001", "execute"), JdbcAgentFake.calls)
    }

    @Test
    fun `applies execution row and fetch limits to jdbc statements`() {
        val agent = createAgent()
        setPrivateConnection(agent, JdbcAgentFake.connection())

        val result = agent.executeQuery(resultSetSql(), null, ExecuteQueryOptions(maxRows = 1, fetchSize = 25))

        assertEquals(listOf(listOf("row-value")), result.rows)
        assertEquals(listOf("setMaxRows:2", "setFetchSize:25", "execute"), JdbcAgentFake.calls)
    }
}
